test/service/test_spreadsheet_generator.py

```diff
@@ -0,0 +1,69 @@
+from broker.service.spreadsheet_generation.spreadsheet_generator import SpreadsheetGenerator, SpreadsheetSpec, TypeSpec, LinkSpec, IncludeSomeModules, IncludeAllModules
+from ingest.api.ingestapi import IngestApi
+
+
+class TestSpreadsheetGenerator(TestCase):
+
+    def test_link_column_generation(self):
+        cell_suspension_spec = TypeSpec("cell_suspension", IncludeAllModules(), False, LinkSpec(["donor_organism"], []))
+
+        ingest_url = "https://api.ingest.dev.archive.data.humancellatlas.org"
+        ingest_api = IngestApi(ingest_url)
+        parsed_tab = SpreadsheetGenerator(ingest_api).tab_for_type(cell_suspension_spec)
+
+        self.assertTrue("donor_organism.biomaterial_core.biomaterial_id" in [col.path for col in parsed_tab.columns])
+
+    def test_context_to_string(self):
+        context = ["project", "project_core", "project_shortname"]
+        self.assertEqual(SpreadsheetGenerator.context_to_path_string(context), "project.project_core.project_shortname")
+
+        context = ["project"]
+        self.assertEqual(SpreadsheetGenerator.context_to_path_string(context), "project")
+
+        context = []
+        self.assertEqual(SpreadsheetGenerator.context_to_path_string(context), "")
+
+    def test_spreadsheet_spec_json_deserialization(self):
+        test_spreadsheet_spec_json = {
+            "types": [{
+                "schemaName": "donor_organism",
+                "includeModules": ["death"],
+                "embedProcess": True,
+                "linkSpec": {
+                    "linkEntities": [],
+                    "linkProtocols": ["biomaterial_collection_protocol"]
+                }}]
+        }
+
+        deserialized_spec = SpreadsheetSpec.from_dict(test_spreadsheet_spec_json)
+        self.assertTrue(len(deserialized_spec.types) == 1)
+        self.assertTrue(deserialized_spec.types[0].schema_name == "donor_organism")
+        self.assertTrue(isinstance(deserialized_spec.types[0].include_modules, IncludeSomeModules))
+        self.assertTrue("death" in deserialized_spec.types[0].include_modules.modules)
+        self.assertTrue(deserialized_spec.types[0].link_spec is not None)
+        self.assertTrue(len(deserialized_spec.types[0].link_spec.link_entities) == 0)
+        self.assertTrue("biomaterial_collection_protocol" in deserialized_spec.types[0].link_spec.link_protocols)
+
+    def test_spreadsheet_spec_hashcode(self):
+        test_type_spec_1 = TypeSpec("project", IncludeSomeModules(["contributors"]), False, LinkSpec(["specimen_from_organism"], []))
+        test_type_spec_2 = TypeSpec("specimen_from_organism", IncludeAllModules(), True, LinkSpec(["donor_organism"], []))
+
+        spreadsheet_spec = SpreadsheetSpec([test_type_spec_1, test_type_spec_2])
+        spreadsheet_spec_dict = spreadsheet_spec.to_dict()
+
+        self.assertEqual(spreadsheet_spec.hashcode(), SpreadsheetSpec.from_dict(spreadsheet_spec_dict).hashcode())
+
+    #@skip
+    def test_generate(self):
```

Comment:
It requires integration with the ingest-api yeah. Kept this here to show Ami the format of an integration test that generates a spreadsheet and then maybe asserts on the column names. I think it can be removed now. Might also look at how to mark tests as integration in `unittest`, I think Ami has written some integ. tests